{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fdf9ccfd-e03c-4bc5-8e60-1ac63d97104f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/maxim/opt/anaconda3/lib/python3.9/site-packages/seaborn/utils.py:10: UserWarning: A NumPy version >=1.22.4 and <2.3.0 is required for this version of SciPy (detected version 1.21.5)\n",
      "  from scipy import stats\n"
     ]
    }
   ],
   "source": [
    "import sims\n",
    "import numpy as np\n",
    "import msprime\n",
    "import pandas as pd\n",
    "\n",
    "import useful\n",
    "import seaborn\n",
    "import matplotlib\n",
    "from matplotlib import pyplot as plt\n",
    "from importlib import reload\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b92b5a25",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5222edba-8556-40ed-b457-5b9721bd67a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ 1896.55172414 18965.51724138  2265.51724138  1034.48275862\n",
      "  2068.96551724  1379.31034483]\n"
     ]
    }
   ],
   "source": [
    "# Generation time, mutation rate and recomination rate\n",
    "RR = 1e-8\n",
    "MU = 1.29e-8 \n",
    "GEN_time = 29.0 \n",
    "\n",
    "# Split Times\n",
    "T_NEAND_migration = 55000 #time of Neanderthal migration into Out_of_africa population\n",
    "T_NEAND_AMH = 550000 # split time between AMH and Neanderthal\n",
    "T_OOF_AF = 65700 # Out_of_Africa migration time\n",
    "T_ND_SPLIT = 60000\n",
    "T_Split_EU_Sample = 40000\n",
    "T_NEAND_samples = 35000\n",
    "T_SECOND_NEAND_MIGRATION = 30000\n",
    "\n",
    "# Effective population size\n",
    "N_ANC = 18500 # N_e of common  AMH and NEanderthal population \n",
    "N_ND = 3400 # N_e of Neanderthal\n",
    "N_AMH = 23000 # N_e of AMH\n",
    "N_OOF = 1861 # N_e of Out_of_Africa population\n",
    "N_AF = 27600 # N_e of Africans\n",
    "N_EU = 13377 #N_e of Europeans\n",
    "N_Sample = 10000\n",
    "\n",
    "N_EU_bottleneck = 1080\n",
    "N_EU_growth = 1450\n",
    "T_EU_growth = 31900\n",
    "gr_rate = 0.00202\n",
    "Portion_admix = 0.1\n",
    "Portion_admix2 = 0.1\n",
    "\n",
    "len_sequence = 2e7 # DNA sequence length\n",
    "\n",
    "n = 50 # number of generated   AF samples\n",
    "n_neand = 6 #number of generated Neanderthals\n",
    "\n",
    "rand_sd =1234 #random seed\n",
    "\n",
    "T = np.array([T_NEAND_migration, T_NEAND_AMH, T_OOF_AF, T_SECOND_NEAND_MIGRATION, T_ND_SPLIT, T_Split_EU_Sample])/GEN_time\n",
    "print(T)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3f3093ce-6e0e-4cc9-95d4-9560763a3b02",
   "metadata": {},
   "outputs": [
    {
     "ename": "InputError",
     "evalue": "Input error in population split: Source and destination populations equal.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mInputError\u001b[0m                                Traceback (most recent call last)",
      "Input \u001b[0;32mIn [3]\u001b[0m, in \u001b[0;36m<cell line: 5>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m N_e \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39marray([N_ANC, N_ND, N_AMH, N_OOF, N_AF, N_EU, N_Sample])\n\u001b[1;32m      4\u001b[0m n_eu\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m100\u001b[39m\n\u001b[0;32m----> 5\u001b[0m ts3400 \u001b[38;5;241m=\u001b[39m\u001b[43msims\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhistory_archaic\u001b[49m\u001b[43m(\u001b[49m\u001b[43mGEN_time\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlen_sequence\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mRR\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mMU\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mN_e\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mT\u001b[49m\u001b[43m,\u001b[49m\u001b[43m  \u001b[49m\u001b[43mn\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrand_sd\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mn_neand\u001b[49m\u001b[43m,\u001b[49m\u001b[43m  \u001b[49m\n\u001b[1;32m      6\u001b[0m \u001b[43m                         \u001b[49m\u001b[43mT_NEAND_samples\u001b[49m\u001b[38;5;241;43m/\u001b[39;49m\u001b[43mGEN_time\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mn_eu\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mN_EU_growth\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m      7\u001b[0m \u001b[43m                         \u001b[49m\u001b[43mT_EU_growth\u001b[49m\u001b[38;5;241;43m/\u001b[39;49m\u001b[43mGEN_time\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mN_EU_bottleneck\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgr_rate\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mPortion_admix\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mPortion_admix2\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      8\u001b[0m sims\u001b[38;5;241m.\u001b[39mprint_neand_dosages(ts3400)\n",
      "File \u001b[0;32m~/PycharmProjects/DAIseg2/sims.py:89\u001b[0m, in \u001b[0;36mhistory_archaic\u001b[0;34m(gen_time, len_seq, rr, mu, n_e, t, n, rand_sd, n_neand, t_neand_samples, n_eu, n_eu_growth, t_eu_growth, n_eu_bottleneck, gr_rt, p_admix, p_admix2)\u001b[0m\n\u001b[1;32m     81\u001b[0m demography\u001b[38;5;241m.\u001b[39madd_admixture(\n\u001b[1;32m     82\u001b[0m     time \u001b[38;5;241m=\u001b[39m t_SECOND_NEAND_MIGRATION,\n\u001b[1;32m     83\u001b[0m     derived \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSAMPLE\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     84\u001b[0m     ancestral \u001b[38;5;241m=\u001b[39m [\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSAMPLE\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mND2\u001b[39m\u001b[38;5;124m\"\u001b[39m],\n\u001b[1;32m     85\u001b[0m     proportions \u001b[38;5;241m=\u001b[39m [\u001b[38;5;241m1\u001b[39m \u001b[38;5;241m-\u001b[39m p_admix2, p_admix2]\n\u001b[1;32m     86\u001b[0m )\n\u001b[1;32m     88\u001b[0m demography\u001b[38;5;241m.\u001b[39msort_events()\n\u001b[0;32m---> 89\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[43mdemography\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdebug\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[1;32m     91\u001b[0m ts \u001b[38;5;241m=\u001b[39m msprime\u001b[38;5;241m.\u001b[39msim_ancestry(\n\u001b[1;32m     92\u001b[0m     samples\u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m     93\u001b[0m         msprime\u001b[38;5;241m.\u001b[39mSampleSet(n_eu, ploidy\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m, population\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mEU\u001b[39m\u001b[38;5;124m'\u001b[39m),\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    102\u001b[0m     record_migrations\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m    103\u001b[0m )\n\u001b[1;32m    105\u001b[0m ts \u001b[38;5;241m=\u001b[39m msprime\u001b[38;5;241m.\u001b[39msim_mutations(ts, rate\u001b[38;5;241m=\u001b[39mmu)\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/msprime/demography.py:1208\u001b[0m, in \u001b[0;36mDemography.debug\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1201\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mdebug\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m   1202\u001b[0m     \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   1203\u001b[0m \u001b[38;5;124;03m    Returns a :class:`.DemographyDebugger` instance for this demography.\u001b[39;00m\n\u001b[1;32m   1204\u001b[0m \n\u001b[1;32m   1205\u001b[0m \u001b[38;5;124;03m    :return: A DemographyDebugger object for this demography.\u001b[39;00m\n\u001b[1;32m   1206\u001b[0m \u001b[38;5;124;03m    :rtype: msprime.DemographyDebugger\u001b[39;00m\n\u001b[1;32m   1207\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 1208\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mDemographyDebugger\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdemography\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/msprime/demography.py:3692\u001b[0m, in \u001b[0;36mDemographyDebugger.__init__\u001b[0;34m(self, Ne, population_configurations, migration_matrix, demographic_events, model, demography)\u001b[0m\n\u001b[1;32m   3690\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdemography \u001b[38;5;241m=\u001b[39m demography\u001b[38;5;241m.\u001b[39mvalidate()\n\u001b[1;32m   3691\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mnum_populations \u001b[38;5;241m=\u001b[39m demography\u001b[38;5;241m.\u001b[39mnum_populations\n\u001b[0;32m-> 3692\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_epochs\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3693\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_misspecification()\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/msprime/demography.py:3697\u001b[0m, in \u001b[0;36mDemographyDebugger._make_epochs\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   3695\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_make_epochs\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m   3696\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mepochs \u001b[38;5;241m=\u001b[39m []\n\u001b[0;32m-> 3697\u001b[0m     simulator \u001b[38;5;241m=\u001b[39m \u001b[43mancestry\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_parse_sim_ancestry\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   3698\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdemography\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdemography\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43minit_for_debugger\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\n\u001b[1;32m   3699\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3700\u001b[0m     start_time \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m\n\u001b[1;32m   3701\u001b[0m     end_time \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/msprime/ancestry.py:1081\u001b[0m, in \u001b[0;36m_parse_sim_ancestry\u001b[0;34m(samples, sequence_length, recombination_rate, gene_conversion_rate, gene_conversion_tract_length, discrete_genome, population_size, demography, ploidy, model, initial_state, start_time, end_time, record_migrations, record_full_arg, additional_nodes, coalescing_segments_only, num_labels, random_seed, init_for_debugger)\u001b[0m\n\u001b[1;32m   1078\u001b[0m random_seed \u001b[38;5;241m=\u001b[39m _parse_random_seed(random_seed)\n\u001b[1;32m   1079\u001b[0m random_generator \u001b[38;5;241m=\u001b[39m _msprime\u001b[38;5;241m.\u001b[39mRandomGenerator(random_seed)\n\u001b[0;32m-> 1081\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mSimulator\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1082\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtables\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minitial_state\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1083\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrecombination_map\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrecombination_map\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1084\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgene_conversion_map\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgene_conversion_map\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1085\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgene_conversion_tract_length\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgene_conversion_tract_length\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1086\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdiscrete_genome\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdiscrete_genome\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1087\u001b[0m \u001b[43m    \u001b[49m\u001b[43mploidy\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mploidy\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1088\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdemography\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdemography\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1089\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmodels\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmodels\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1090\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstore_migrations\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrecord_migrations\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1091\u001b[0m \u001b[43m    \u001b[49m\u001b[43madditional_nodes\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43madditional_nodes\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1092\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcoalescing_segments_only\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcoalescing_segments_only\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1093\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstart_time\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstart_time\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1094\u001b[0m \u001b[43m    \u001b[49m\u001b[43mend_time\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mend_time\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1095\u001b[0m \u001b[43m    \u001b[49m\u001b[43mnum_labels\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnum_labels\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1096\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrandom_generator\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrandom_generator\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1097\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/msprime/ancestry.py:1426\u001b[0m, in \u001b[0;36mSimulator.__init__\u001b[0;34m(self, tables, recombination_map, gene_conversion_map, gene_conversion_tract_length, discrete_genome, ploidy, demography, random_generator, models, store_migrations, additional_nodes, coalescing_segments_only, start_time, end_time, num_labels)\u001b[0m\n\u001b[1;32m   1423\u001b[0m gene_conversion_rate \u001b[38;5;241m=\u001b[39m gene_conversion_map\u001b[38;5;241m.\u001b[39mrate[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m   1425\u001b[0m start_time \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m start_time \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;28;01melse\u001b[39;00m start_time\n\u001b[0;32m-> 1426\u001b[0m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;21;43m__init__\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1427\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtables\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mll_tables\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1428\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrecombination_map\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mll_recomb_map\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1429\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstart_time\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstart_time\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1430\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrandom_generator\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrandom_generator\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1431\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmigration_matrix\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdemography\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmigration_matrix\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1432\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpopulation_configuration\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mll_population_configuration\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1433\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdemographic_events\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mll_demographic_events\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1434\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstore_migrations\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstore_migrations\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1435\u001b[0m \u001b[43m    \u001b[49m\u001b[43madditional_nodes\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43madditional_nodes\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvalue\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1436\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcoalescing_segments_only\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcoalescing_segments_only\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1437\u001b[0m \u001b[43m    \u001b[49m\u001b[43mnum_labels\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnum_labels\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1438\u001b[0m \u001b[43m    \u001b[49m\u001b[43msegment_block_size\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msegment_block_size\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1439\u001b[0m \u001b[43m    \u001b[49m\u001b[43mavl_node_block_size\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mavl_node_block_size\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1440\u001b[0m \u001b[43m    \u001b[49m\u001b[43mnode_mapping_block_size\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnode_mapping_block_size\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1441\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgene_conversion_rate\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgene_conversion_rate\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1442\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgene_conversion_tract_length\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgene_conversion_tract_length\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1443\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdiscrete_genome\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdiscrete_genome\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1444\u001b[0m \u001b[43m    \u001b[49m\u001b[43mploidy\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mploidy\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1445\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1446\u001b[0m \u001b[38;5;66;03m# Highlevel attributes used externally that have no lowlevel equivalent\u001b[39;00m\n\u001b[1;32m   1447\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mend_time \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39minf \u001b[38;5;28;01mif\u001b[39;00m end_time \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;28;01melse\u001b[39;00m end_time\n",
      "\u001b[0;31mInputError\u001b[0m: Input error in population split: Source and destination populations equal."
     ]
    }
   ],
   "source": [
    "N_ND = 3400 # N_e of Neanderthal\n",
    "N_e = np.array([N_ANC, N_ND, N_AMH, N_OOF, N_AF, N_EU, N_Sample])\n",
    "\n",
    "n_eu=100\n",
    "ts3400 =sims.history_archaic(GEN_time, len_sequence, RR, MU, N_e, T,  n, rand_sd, n_neand,  \n",
    "                         T_NEAND_samples/GEN_time, n_eu, N_EU_growth, \n",
    "                         T_EU_growth/GEN_time, N_EU_bottleneck, gr_rate, Portion_admix, Portion_admix2)\n",
    "sims.print_neand_dosages(ts3400)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a3fe960-36b6-4449-b708-834814adfe50",
   "metadata": {},
   "outputs": [],
   "source": [
    "ND_true_tracts = [[],[]]\n",
    "for idx in range(0, n_eu): \n",
    "#    print( sims.get_migrating_tracts_ind(ts3400, 'NEAND', idx, T[0]))\n",
    "    ND_true_tracts[0].append( sims.get_migrating_tracts_ind(ts3400, 'NEAND', idx, T[0]) )\n",
    "    ND_true_tracts[1].append( sims.get_migrating_tracts_ind(ts3400, 'NEAND', idx, T[3]) )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99fa83a9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdbb589b-810c-499d-8ba3-054cc1d81ec0",
   "metadata": {},
   "outputs": [],
   "source": [
    "N_ref_pop=50\n",
    "N_neanderthal=4\n",
    "N=3\n",
    "with open('make.obs.sims.txt', 'w') as f:\n",
    "    f.write('#POSITIONS\\t#REF\\t#ALT\\tANCESTRAL\\t#OUTGROUP\\t#ARCHAIC\\t#OBSERVATIONS\\n')\n",
    "    for v in ts3400.variants():\n",
    "        outgroup= str(list(set(v.genotypes[n_eu :( n_eu+N_ref_pop)]))).replace('[','').replace(']','').replace(' ','')\n",
    "        archaic= str(list(set(v.genotypes[n_eu+n :( n_eu+n+N_neanderthal)]))).replace('[','').replace(']','').replace(' ','')\n",
    "\n",
    "        obs=''\n",
    "        for i in v.genotypes[0 :n_eu]:\n",
    "            obs+=str(i)+' '\n",
    "\n",
    "\n",
    "        f.write(str(int(v.site.position))+'\\t'+str(v.alleles[0])+'\\t'+\n",
    "                str(v.alleles[1]) + '\\t'+ str(0)+'\\t' + outgroup+'\\t'+archaic+'\\t'+str(obs)+'\\n')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09310ca3-8ceb-4d2a-89d6-bd8c2829b80e",
   "metadata": {},
   "outputs": [],
   "source": [
    "L=1000\n",
    "\n",
    "#create bed file\n",
    "with open('test.bed','w') as f:\n",
    "    f.write('1\\t0\\t'+str(int(len_sequence)-1)+'\\n')\n",
    "\n",
    "\n",
    "\n",
    "f_bed='test.bed'\n",
    "domain=useful.read_bed(f_bed)\n",
    "\n",
    "n_windows=(domain[-1][1]-domain[0][0])//L + 1\n",
    "windows_cover=np.ones(n_windows)*0.999\n",
    "\n",
    "#create archaic covering file. \n",
    "CHR=1\n",
    "with open('arch.covering.chr'+str(CHR)+'.txt','w') as f:\n",
    "    for j in windows_cover:\n",
    "        f.write(str(j)+'\\n')\n",
    "\n",
    "print(n_eu)\n",
    "#create file with sample's names\n",
    "with open('samples.txt','w') as f:\n",
    "    for i in range(int(n_eu/2)):\n",
    "        f.write('eu'+str(i)+'\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "434eb401-e4bf-4d4f-801d-5d6d7db86aff",
   "metadata": {},
   "outputs": [],
   "source": [
    "! python3 dai.seg.py --obs_samples samples.txt --bed test.bed   --HMM_par par.file.txt --EM no --prepared_file make.obs.sims.txt --o out --arch_cover arch.covering.chr1.txt --cut_off 0.9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85aff461-71a7-4983-a519-ebfca71c0dc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "reload(useful)\n",
    "HMM_tracts, REAL_tracts=[], []\n",
    "HMM_ND1, HMM_ND2=useful.read_out('out.archaic.ND1.txt'), useful.read_out('out.archaic.ND2.txt')\n",
    "for j in range(len(HMM_ND1)):\n",
    "\n",
    "    nd_hmm=HMM_ND1[j]+HMM_ND2[j]\n",
    "    if [] in nd_hmm:\n",
    "        nd_hmm.remove([])\n",
    "    modern_hmm=useful.tracts_eu(nd_hmm, len_sequence)\n",
    "    HMM_tracts.append([modern_hmm, HMM_ND1[j], HMM_ND2[j]])\n",
    "    \n",
    "    \n",
    "    nd_real=ND_true_tracts[0][j]+ND_true_tracts[1][j]    \n",
    "    if [] in nd_real:\n",
    "        nd_real.remove([])\n",
    "        \n",
    "    modern_real=useful.tracts_eu(nd_real, len_sequence)\n",
    "    REAL_tracts.append([modern_real, ND_true_tracts[0][j], ND_true_tracts[1][j]])    \n",
    "\n",
    "\n",
    "\n",
    "useful.confusion_mtrx(REAL_tracts[0], HMM_tracts[0])   \n",
    "df = useful.df_result(  REAL_tracts, HMM_tracts,N_neanderthal, L, N_ref_pop, n_eu, N)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8d2759c-dfaa-4337-ba6c-c0fb1865a20d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8964969b-6a79-4fa9-aa6e-e317dbd9a735",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cec0c9fc-fc40-4c63-8d6e-b841587ba930",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e7173ee-3337-415b-952d-f290f537a4cd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36379b24-d87d-4c2d-87fc-df7bc9ec7256",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_chunk_lengths(intervals):\n",
    "    if not isinstance(intervals, list):\n",
    "        return []\n",
    "    lengths = []\n",
    "    for element in intervals:\n",
    "        if isinstance(element, list) and len(element) == 2 and isinstance(element[0], (int, float)):\n",
    "            lengths.append(element[1] - element[0])\n",
    "        elif isinstance(element, list):\n",
    "            lengths.extend(compute_chunk_lengths(element))\n",
    "    return lengths\n",
    "\n",
    "chunk_lengths1 = compute_chunk_lengths(HMM_ND1)\n",
    "chunk_lengths2 = compute_chunk_lengths(HMM_ND2)\n",
    "\n",
    "plt.figure(figsize=(8, 5))\n",
    "plt.hist(chunk_lengths1, bins=50, alpha=0.7, color='blue')\n",
    "plt.xlabel(\"Длина участка (bp)\")\n",
    "plt.ylabel(\"Количество\")\n",
    "plt.title(\"Распределение длин неандертальских участков первого примешивания\")\n",
    "plt.show()\n",
    "\n",
    "plt.figure(figsize=(8, 5))\n",
    "plt.hist(chunk_lengths2, bins=50, alpha=0.7, color='red')\n",
    "plt.xlabel(\"Длина участка (bp)\")\n",
    "plt.ylabel(\"Количество\")\n",
    "plt.title(\"Распределение длин неандертальских участков второго примешивания\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8defcb8c-17a6-49de-9beb-393f2d476543",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c56d488d-caa5-45d8-a5be-fd2143a3613f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge(intervals):\n",
    "    new_intervals = []\n",
    "    \n",
    "    for element in intervals:\n",
    "        if isinstance(element, list) and len(element) == 2 and isinstance(element[0], (int, float)):\n",
    "            new_intervals.append(element)\n",
    "        elif isinstance(element, list):\n",
    "            new_intervals.extend(merge(element))\n",
    "    new_intervals.sort()\n",
    "    merged = []\n",
    "    for start, end in new_intervals:\n",
    "        if not merged or merged[-1][1] < start:  \n",
    "            merged.append([start, end])\n",
    "        else:\n",
    "            merged[-1][1] = max(merged[-1][1], end)\n",
    "\n",
    "    return merged\n",
    "\n",
    "def inter_len(intervals):\n",
    "    return sum(end - start for start, end in intervals)\n",
    "\n",
    "def inter_diff(a, b):\n",
    "    return subtract(union(a, b), intersect(a, b))\n",
    "\n",
    "def intersect(a, b):\n",
    "    i, j = 0, 0\n",
    "    res = []\n",
    "    while i < len(a) and j < len(b):\n",
    "        startA, endA = a[i]\n",
    "        startB, endB = b[j]\n",
    "        if endA < startB:\n",
    "            i += 1\n",
    "        elif endB < startA:\n",
    "            j += 1\n",
    "        else:\n",
    "            left = max(startA, startB)\n",
    "            right = min(endA, endB)\n",
    "            if left < right:\n",
    "                res.append([left, right])\n",
    "            if endA < endB:\n",
    "                i += 1\n",
    "            else:\n",
    "                j += 1\n",
    "    return res\n",
    "\n",
    "def union(a, b):\n",
    "    combined = merge(a + b)\n",
    "    return combined\n",
    "\n",
    "def subtract(a, b):\n",
    "    res = []\n",
    "    i, j = 0, 0\n",
    "    while i < len(a):\n",
    "        startA, endA = a[i]\n",
    "        if j >= len(b):\n",
    "            res.append([startA, endA])\n",
    "            i += 1\n",
    "            continue\n",
    "\n",
    "        startB, endB = b[j]\n",
    "        if endA < startB:\n",
    "            res.append([startA, endA])\n",
    "            i += 1\n",
    "        elif endB < startA:\n",
    "            j += 1\n",
    "        else:\n",
    "            if startA < startB:\n",
    "                res.append([startA, startB])\n",
    "            if endA <= endB:\n",
    "                i += 1\n",
    "            else:\n",
    "                a[i] = [endB, endA]\n",
    "                j += 1\n",
    "    return res\n",
    "\n",
    "true_intervals = ND_true_tracts[0]\n",
    "pred_intervals = HMM_ND1\n",
    "\n",
    "def one_dim_intervals(inter):\n",
    "    result = []\n",
    "    for x in inter:\n",
    "        if isinstance(x, list) and len(x) == 2 and isinstance(x[0], (int, float)):\n",
    "            result.append(x)\n",
    "        elif isinstance(x, list):\n",
    "            result.extend(one_dim_intervals(x))\n",
    "    return result\n",
    "\n",
    "true_flat = one_dim_intervals(true_intervals)\n",
    "pred_flat = one_dim_intervals(pred_intervals)\n",
    "true_merged = merge(true_flat)\n",
    "pred_merged = merge(pred_flat)\n",
    "\n",
    "mismatch_intervals = inter_diff(true_merged, pred_merged)\n",
    "mismatch_length = inter_len(mismatch_intervals)\n",
    "\n",
    "sum_true = inter_len(true_merged)\n",
    "if sum_true > 0:\n",
    "    mismatch_rate = mismatch_length / sum_true\n",
    "    \n",
    "print(\"Число отличий:\", mismatch_length)\n",
    "print(\"Общая длина:\", sum_true)\n",
    "print(\"Отличий на единицу длины:\", mismatch_rate)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12d6cb61",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
